torch/utils/mkldnn.py

```diff
@@ -0,0 +1,32 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import torch
+
+
+def to_mkldnn(module):
+    def t_fn(t):
+        if t.is_floating_point():
+            return t.to_mkldnn()
+
+    def m_fn(m):
+        for param in m._parameters.values():
+            if param is not None:
+                # Tensors stored in modules are graph leaves, and we don't
+                # want to create copy nodes, so we have to unpack the data.
+                param.data = t_fn(param.data)
+                if param._grad is not None:
+                    param._grad.data = t_fn(param._grad.data)
+
+        for key, buf in m._buffers.items():
+            if buf is not None:
+                m._buffers[key] = t_fn(buf)
+
+        if isinstance(m, torch.nn.Conv2d):
+            m.weight.data = torch._C._nn.mkldnn_reorder_conv2d_weight(
```

Comment:
we will probably need to move this call elsewhere in order to define "prepack for inference" interface in the future. Otherwise it's hard to serialize the mkldnn including the particular layout in the file. But it can be done in a separate PR when we get to more API pieces

Reply:
> we will probably need to move this call elsewhere in order to define "prepack for inference" interface in the future. Otherwise it's hard to serialize the mkldnn including the particular layout in the file.

The pre-packing is also useful for training since it avoids extra reorders as well. In order to serialize the weights properly, can we always `to_dense` when the module is serialized?

Reply:
For training more work would be needed - for example handling mkldnn tensors in optimizer

What I meant about inference is that we'd need a mechanism to bundle instructions to pack the weights after deserialization to mkldnn format (since as you said they'd be stored in as regular to_dense tensors). We have a prototype mechanism for it (https://github.com/pytorch/pytorch/blob/master/torch/jit/quantized.py#L34) but it needs to be improved. Most likely, we'd need to swap out Conv2d module with special MkldnnConv2d module that handles packing after loading